CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/LayersViewPane.java

```diff
@@ -109,47 +115,101 @@ public LayersViewPane(final LayersViewController controller) {
         options.prefWidthProperty().bind(layersViewPane.widthProperty());
 
         this.setCenter(layersViewPane);
+        controller.writeState();
     }
 
-    private void createLayer(final boolean defaultLayer) {
-        final Label layerIdText = new Label(String.format("%02d", ++currentIndex));
+    public LayersViewController getController() {
+        return controller;
+    }
+
+    private int createLayer(final int currentIndex, final boolean checkBoxSelected, final String query, final String description) {
+        final Label layerIdText = new Label(String.format("%02d", currentIndex));
         layerIdText.setMinWidth(30);
         layerIdText.setPrefWidth(40);
         layerIdText.setTextAlignment(TextAlignment.CENTER);
         layerIdText.setPadding(new Insets(0, 0, 0, 10));
 
-        final CheckBox visibilityCheckBox = new CheckBox();
-        visibilityCheckBox.setMinWidth(60);
-        visibilityCheckBox.setPadding(new Insets(0, 30, 0, 15));
-        visibilityCheckBox.selectedProperty().addListener((observable, oldValue, newValue) -> {
+        final Node visibilityCheckBox = new CheckBox();
+        ((CheckBox) visibilityCheckBox).setMinWidth(60);
+        ((CheckBox) visibilityCheckBox).setPadding(new Insets(0, 30, 0, 15));
+        ((CheckBox) visibilityCheckBox).setSelected(checkBoxSelected);
+        visibilityCheckBox.setOnMouseClicked(e -> {
+            final Node source = (Node) e.getSource();
+            final LayerDescription layer = layers.get(GridPane.getRowIndex(source) - 1);
+            layer.setCurrentLayerVisibility(!layer.getCurrentLayerVisibility());
             controller.submit();
             controller.execute();
+            controller.writeState();
         });
 
-        final TextArea queryTextArea = new TextArea();
-        queryTextArea.setPrefRowCount(1);
+        final Node queryTextArea = new TextArea();
+        ((TextArea) queryTextArea).setPrefRowCount(1);
+        ((TextArea) queryTextArea).setText(query);
+        ((TextArea) queryTextArea).focusedProperty().addListener((observable, oldVal, newVal) -> {
+            if (!newVal) {
+                final LayerDescription layer = layers.get(currentIndex - 1);
+                layer.setQueryText(((TextArea) queryTextArea).getText());
+                controller.writeState();
+            }
+        });
 
-        final TextArea descriptionTextArea = new TextArea();
-        descriptionTextArea.setPrefRowCount(1);
+        final Node descriptionTextArea = new TextArea();
```

Comment:
Curious as to why this is declared as a Node rather than a TextArea (given all the subsequent typecasting)?

Reply:
I couldn't manage to get an event to trigger on the `TextArea` when it was inserted in the `GridPane`. So I had to declare it as a Node, then add the event to it.